puppet/zulip: Replace apt::source and apt::ppa with setup-apt-repo.

rht opened the pull request:
<!-- What's this PR for?  (Just a link to an issue is fine.) -->
The remaining `apt::*` are located in `zulip_ops` though:
1. The apt sources are not covered by setup-apt-repo.
```
manifests/zmirror_personals.pp
22:  apt::source {'debathena':

manifests/zmirror.pp
22:  apt::source {'debathena':
```
2. They are not required for general production env anyway.

Once they have been moved to a script, it should be possible to remove all the apt and stdlib modules. And this will result with the puppet setup becoming distribution-independent. The remaining step for the general production env to be distribution independent is to port `setup-apt-repo` to other distributions.

**Testing Plan:** <!-- How have you tested? -->


**GIFs or Screenshots:** <!-- If a UI change.  See:
  https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
  -->


<!-- Also be sure to make clear, coherent commits:
  https://zulip.readthedocs.io/en/latest/contributing/version-control.html
  -->

timabbott commented on puppet/zulip/manifests/postgres_appdb_base.pp:
```diff
@@ -8,11 +9,6 @@
-    $appdb_packages:
-      ensure  => 'installed',
-      require => Apt::Source['zulip'],
-  }
```
This still needs e.g. `require => Exec['setup_apt_repo']`

timabbott commented on puppet/zulip/manifests/postgres_appdb_base.pp:
```diff
@@ -49,10 +45,6 @@
   if $pgroonga == 'enabled' {
-    apt::ppa {'ppa:groonga/ppa':
-      before => Package["postgresql-${zulip::base::postgres_version}-pgroonga"],
-    }
```
Similarly here, I think -- we still want to depend on `setup-apt-repo` being run.

timabbott commented on puppet/zulip/manifests/apt_repository.pp:
```diff
@@ -1,24 +1,11 @@
-      include_src => true,
-    }
+  #$prefix = "/home/zulip/deployments/current/scripts"
+  $setup_apt_repo_file = "../scripts/lib/setup-apt-repo"
```
This is sorta an unpleasant issue; I think we may need to pass the path in via e.g. an environment variable in `zulip-puppet-apply`.

timabbott commented on puppet/zulip/manifests/apt_repository.pp:
```diff
@@ -1,24 +1,11 @@
+  exec{'setup_apt_repo':
+    require => File[$setup_apt_repo_file],
+    command => "bash -c '${setup_apt_repo_file} && touch ${setup_applied_file}'",
+    creates => "${setup_applied_file}",
```
I think we might want to just optimize `./scripts/lib/setup-apt-repo` and then run it unconditionally... otherwise, we'll lose the feature of "if we need to rotate an apt key, upgrades work" with this change.

timabbott commented on scripts/lib/setup-apt-repo-debathena:
```diff
@@ -0,0 +1,33 @@
+    # `apt-get update` finishes successfully.
+    touch "$STAMP_FILE"
+    apt-get update && rm -f "$STAMP_FILE"
+fi
```
So, there's a preparatory project I'd like to do before we copy this file, which is to add `set -e` to `setup-apt-repo`.  

Also, the copying is probably wrong, in that `SOURCES_FILE` and `STAMP_FILE` should presumably be renamed to have `debathena` in their name (or something).

timabbott replied:
(Getting `set -e` working is a bit of work, since we need to add handling for e.g. `$SOURCES_FILE` not existing)

timabbott replied:
Sounds good.

timabbott commented on puppet/zulip/manifests/apt_repository.pp:
```diff
@@ -1,24 +1,8 @@
-    }
+  $setup_apt_repo_file = "${::zulip_scripts_path}/lib/setup-apt-repo"
+  exec{'setup_apt_repo':
+    # require => File[$setup_apt_repo_file],
```
I think this comment should just be deleted, right?

timabbott commented:
Maybe we can grep for all the functions defined in `puppet/stdlib/spec/unit/puppet/parser/functions` or something.  I have this vague recollection of using something from it recently, maybe in 5e7aa27c291ec1153f9e22f2af902cbe68d8b1b7?  

I suppose we could scope this PR for now to just removing `apt`; for `stdlib`, I think we'll want to find the ~5 functions we use and keep those (but delete the rest).

rht commented:
The stdlib is indeed used in lots of places:
```
$ ls -1 stdlib/lib/puppet/parser/functions | rev | cut -c4- | rev | xargs -I _ rg -l '_\(' | grep -v stdlib
zulip/files/nagios_plugins/zulip_postgres_common/check_postgres.pl
zulip/files/nagios_plugins/zulip_postgres_common/check_postgres.pl
zulip/files/nagios_plugins/zulip_postgres_appdb/check_fts_update_log
zulip/files/nagios_plugins/zulip_postgres_common/check_postgres.pl
zulip/files/postgresql/process_fts_updates
zulip_ops/files/zulip-ec2-configure-interfaces
zulip_ops/manifests/nagios.pp
zulip/files/postgresql/pg_backup_and_purge
zulip/files/postgresql/process_fts_updates
zulip/templates/supervisor/zulip.conf.template.erb
zulip_ops/files/nagios_plugins/zulip_zephyr_mirror/check_user_zephyr_mirror_liveness
zulip_ops/files/nagios_plugins/zulip_zephyr_mirror/check_personal_zephyr_mirrors
zulip/files/nagios_plugins/zulip_app_frontend/check_queue_worker_errors
zulip/files/nagios_plugins/zulip_postgres_common/check_postgres.pl
zulip/lib/puppet/parser/functions/zulipconf.rb
zulip/lib/puppet/parser/functions/zulipsecret.rb
zulip/files/postgresql/pg_backup_and_purge
zulip/files/nagios_plugins/zulip_base/check_debian_packages
zulip/files/nagios_plugins/zulip_app_frontend/check_send_receive_time
zulip_ops/files/zulip-ec2-configure-interfaces
zulip/manifests/app_frontend_base.pp
zulip/files/nagios_plugins/zulip_postgres_common/check_postgres.pl
zulip/files/nagios_plugins/zulip_postgres_common/check_postgres.pl
zulip/templates/postgresql/9.6/postgresql.conf.template.erb
zulip/templates/postgresql/9.5/postgresql.conf.template.erb
zulip/templates/postgresql/10/postgresql.conf.template.erb
zulip/templates/postgresql/9.3/postgresql.conf.template.erb
zulip_ops/files/nagios_plugins/zulip_zephyr_mirror/check_zephyr_mirror
zulip_ops/files/nagios_plugins/zulip_zephyr_mirror/check_user_zephyr_mirror_liveness
zulip_ops/files/nagios_plugins/zulip_zephyr_mirror/check_personal_zephyr_mirrors
zulip_ops/files/zulip-ec2-configure-interfaces
zulip/files/postgresql/pg_backup_and_purge
zulip_ops/files/nagios_plugins/zulip_zephyr_mirror/check_zephyr_mirror
zulip_ops/files/nagios_plugins/zulip_zephyr_mirror/check_personal_zephyr_mirrors
zulip/files/nagios_plugins/zulip_postgres_common/check_postgres_backup
zulip/files/nagios_plugins/zulip_app_frontend/check_cron_file
zulip/lib/puppet/parser/functions/zulipconf.rb
zulip/lib/puppet/parser/functions/zulipsecret.rb
zulip/files/nagios_plugins/zulip_app_frontend/check_send_receive_time
zulip/files/nagios_plugins/zulip_postgres_common/check_postgres.pl
zulip/files/nagios_plugins/zulip_app_frontend/check_cron_file
zulip_ops/files/nagios_plugins/zulip_zephyr_mirror/check_zephyr_mirror
zulip_ops/files/nagios_plugins/zulip_zephyr_mirror/check_user_zephyr_mirror_liveness
zulip_ops/files/nagios_plugins/zulip_zephyr_mirror/check_personal_zephyr_mirrors
zulip/templates/postgresql/9.6/postgresql.conf.template.erb
zulip/templates/postgresql/10/postgresql.conf.template.erb
zulip/templates/postgresql/9.3/postgresql.conf.template.erb
zulip/templates/postgresql/9.5/postgresql.conf.template.erb
zulip/lib/puppet/type/chroot.rb
zulip/lib/puppet/type/apache2site.rb
zulip/lib/puppet/type/apache2mod.rb
zulip_ops/files/zulip-ec2-configure-interfaces
```

I have amended the last commit to only remove puppetlabs-apt.

timabbott commented:
I think your grep is off; I don't think all of those files actually use things from the puppet stdlib.  For one thing, many of them aren't even `.pp` type files.

rht commented:
Yeah I found lots of false positives such as `.strip()`, `validate_range()`.

rht commented:
Ok, found only 2 with a much better filetype filter (with the erubis files ignored):
```
ls -1 stdlib/lib/puppet/parser/functions | rev | cut -c4- | rev | xargs -I {} rg -tpuppet -Truby -g '!stdlib/*' '{}\('  zulip_ops/manifests/nagios.pp
12:  $nagios_format_users = join($zulip_ops::base::users, ',')

zulip/templates/supervisor/zulip.conf.template.erb
76:command=/home/zulip/deployments/current/manage.py process_queue --multi_threaded <%= @queues.join(' ') %>
zulip/manifests/app_frontend_base.pp
61:    $tornado_ports = range(9800, 9800 + $tornado_processes)
zulip/templates/postgresql/9.5/postgresql.conf.template.erb
336:                                    # can include strftime() escapes

zulip/templates/postgresql/9.6/postgresql.conf.template.erb
347:                                    # can include strftime() escapes

zulip/templates/postgresql/10/postgresql.conf.template.erb
359:                                    # can include strftime() escapes

zulip/templates/postgresql/9.3/postgresql.conf.template.erb
291:                                    # can include strftime() escapes
zulip/templates/postgresql/9.5/postgresql.conf.template.erb
336:                                    # can include strftime() escapes

zulip/templates/postgresql/9.6/postgresql.conf.template.erb
347:                                    # can include strftime() escapes

zulip/templates/postgresql/9.3/postgresql.conf.template.erb
291:                                    # can include strftime() escapes

zulip/templates/postgresql/10/postgresql.conf.template.erb
359:                                    # can include strftime() escapes
```

rht commented:
Or with '*.pp' only filter,
``
ls -1 stdlib/lib/puppet/parser/functions | rev | cut -c4- | rev | xargs -I {} rg -g '*.pp' '{}\('  
stdlib/tests/has_interface_with.pp
2:info('has_interface_with(\'lo\'):', has_interface_with('lo'))
3:info('has_interface_with(\'loX\'):', has_interface_with('loX'))
4:info('has_interface_with(\'ipaddress\', \'127.0.0.1\'):', has_interface_with('ipaddress', '127.0.0.1'))
5:info('has_interface_with(\'ipaddress\', \'127.0.0.100\'):', has_interface_with('ipaddress', '127.0.0.100'))
6:info('has_interface_with(\'network\', \'127.0.0.0\'):', has_interface_with('network', '127.0.0.0'))
7:info('has_interface_with(\'network\', \'128.0.0.0\'):', has_interface_with('network', '128.0.0.0'))
8:info('has_interface_with(\'netmask\', \'255.0.0.0\'):', has_interface_with('netmask', '255.0.0.0'))
9:info('has_interface_with(\'netmask\', \'256.0.0.0\'):', has_interface_with('netmask', '256.0.0.0'))
stdlib/tests/has_ip_address.pp
2:info('has_ip_address(\'192.168.1.256\'):', has_ip_address('192.168.1.256'))
3:info('has_ip_address(\'127.0.0.1\'):', has_ip_address('127.0.0.1'))
stdlib/tests/has_ip_network.pp
2:info('has_ip_network(\'127.0.0.0\'):', has_ip_network('127.0.0.0'))
3:info('has_ip_network(\'128.0.0.0\'):', has_ip_network('128.0.0.0'))
zulip_ops/manifests/nagios.pp
12:  $nagios_format_users = join($zulip_ops::base::users, ',')
zulip/manifests/app_frontend_base.pp
61:    $tornado_ports = range(9800, 9800 + $tornado_processes)
```

It is safe to conclude there are only 2 stdlib functions being used, though I am not sure why this is needed beyond what can be done in Ruby, `$zulip_ops::base::users.join(',')`, `9800..9800+$tornado_processes`.

timabbott commented:
Can we try just copying those two functions into `puppet/zulip/lib/puppet/` and then kill the library?

rht commented:
OKk I just did that in the last 2 commits.

timabbott commented:
A few things:
* Can you update `docs/THIRDPARTY` and linters in the appropriate places they mention these modules and the functions we've copied?
* `setup-apt-repo-debathena` should use a different `SOURCES_FILE`.
* Have you tested that the `join` functions work in this edited state?

With that, we're just blocked on `scripts/setup/setup-apt-repo` perf

rht commented:
> Can you update docs/THIRDPARTY and linters in the appropriate places they mention these modules and the functions we've copied?

Updated docs/THIRDPARTY. I added the linter modification in a separate commit.

> setup-apt-repo-debathena should use a different SOURCES_FILE.

Renamed zulip.list -> zulip_debathena.list

> Have you tested that the join functions work in this edited state?

Not yet, but then, removing stdlib can be decoupled from removing apt.

rht commented:
I might have written this somewhere else, but one idea to speed up `setup-apt-repo` is to do the apt-get install gpg https-transport etc only when it has been confirmed that zulip.list and zulip_debathena.list do not contain the sources. This can be checked by either doing a string match or maybe a sha1sum of the file.

rht commented:
Though the sources check might skip through the apt-get update if it failed previously due to network problem.

rht commented:
Posted another solution of not having to run the bulk of setup-apt-repo at the last commit.

timabbott commented on scripts/lib/setup-apt-repo:
```diff
@@ -4,6 +4,12 @@ set -e
 
 SOURCES_FILE=/etc/apt/sources.list.d/zulip.list
 STAMP_FILE=/etc/apt/sources.list.d/zulip.list.apt-update-in-progress
+SUCCESS_FILE=/etc/apt/sources.list.d/zulip.list.success
```
This doesn't solve the "files changed" problem.  As I said, I think we want to hash this file + `scripts/setup/*.asc`

rht commented:
The diff of the join.rb contains only comments and a different error message for wrong number of arg.
```
$ diff old_join.rb new_join.rb
> # Taken from https://github.com/puppetlabs/puppetlabs-stdlib/blob/19cdf29f27c3e5005ee441d1ec46d7da27a0f777/lib/puppet/parser/functions/join.rb
4d4
< 
6,7c6,9
<   newfunction(:join, :type => :rvalue, :doc => <<-EOS
< This function joins an array into a string using a seperator.
---
>   newfunction(:join, :type => :rvalue, :doc => <<-DOC
>     This function joins an array into a string using a separator.
> 
>     *Examples:*
9c11
< *Examples:*
---
>         join(['a','b','c'], ",")
11c13
<     join(['a','b','c'], ",")
---
>     Would result in: "a,b,c"
13,15c15,18
< Would result in: "a,b,c"
<     EOS
<   ) do |arguments|
---
>     Note: from Puppet 5.4.0, the compatible function with the same name in Puppet core
>     will be used instead of this function.
>     DOC
>              ) do |arguments|
18,19c21
<     raise(Puppet::ParseError, "join(): Wrong number of arguments " +
<       "given (#{arguments.size} for 1)") if arguments.size < 1
---
>     raise(Puppet::ParseError, "join(): Wrong number of arguments given (#{arguments.size} for 1)") if arguments.empty?
```

rht commented:
@timabbott I pushed another idea in the new last commit.

rht commented on scripts/lib/setup-apt-repo:
```diff
@@ -42,6 +35,18 @@ fi
+        apt-key add "$SCRIPTS_PATH"/setup/pgroonga-ppa.asc
+        apt-key add "$SCRIPTS_PATH"/setup/zulip-ppa.asc
+    elif [ "$release" = "stretch" ]; then
+        apt-get install -y lsb-release apt-transport-https gnupg debian-archive-keyring
```
This is so that `apt-get` is run only once. Previously it was `apt-get install -y lsb-release apt-transport-https gnupg` and `apt-get install -y debian-archive-keyring`.

rht commented:
@timabbott the no-op for setup-apt-repo in the last commit should be pretty fast.

timabbott commented:
OK, I merged 0d48f47eba03fe04d0ce37e2e849733f804e6549 with a version that I think is actually reasonable.

timabbott commented:
Can you rework the `.debathena` one to be based on the newly merged script?

Once that's done, we can try to integrate the rest.

rht commented:
Yeah I just did that (basing the debathena version on https://github.com/zulip/zulip/commit/0d48f47eba03fe04d0ce37e2e849733f804e6549) and rebased.

timabbott commented:
OK, I merged this, after renaming zulip_ops::apt_repository to zulip_ops::apt_repository_debathena for clarity, and editing the last commit message to close #7423.  Thanks @rht!
